src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -0,0 +1,488 @@
+
+                // At this point, we have a valid two-byte start of a four-byte sequence.
+
+                if (data.Length < 3)
+                {
+                    // Two bytes of an incomplete four-byte sequence.
+                    goto TwoBytesOfIncompleteMultiByteSequence;
+                }
+                else
+                {
+                    byte thirdByte = data[2];
+                    if (IsUtf8ContinuationByte(thirdByte))
+                    {
+                        // Valid three-byte start of a four-byte sequence.
+
+                        if (data.Length < 4)
+                        {
+                            // Three bytes of an incomplete four-byte sequence.
+                            goto ThreeBytesOfIncompleteMultiByteSequence;
+                        }
+                        else
+                        {
+                            byte fourthByte = data[3];
+                            if (IsUtf8ContinuationByte(fourthByte))
+                            {
+                                // Well-formed four-byte sequence.
+                                scalar |= (((uint)thirdByte & 0x3FU) << 6) | ((uint)fourthByte & 0x3FU);
+                                scalarValue = UnicodeScalar.CreateWithoutValidation(scalar);
+                                numBytesConsumed = 4;
+                                return SequenceValidity.WellFormed;
+                            }
+                            else
+                            {
+                                // Three bytes of an improperly terminated multi-byte sequence.
+                                goto InvalidThreeByteSequence;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // Two bytes of improperly terminated multi-byte sequence.
+                        goto InvalidTwoByteSequence;
+                    }
+                }
+            }
+
+        // Everything below here is error handling.
+
+        InvalidOneByteSequence:
+            numBytesConsumed = 1;
+            return SequenceValidity.Invalid;
+
+        InvalidTwoByteSequence:
+        OverlongOutOfRangeOrSurrogateSequence:
+            numBytesConsumed = 2;
+            return SequenceValidity.Invalid;
+
+        InvalidThreeByteSequence:
+            numBytesConsumed = 3;
+            return SequenceValidity.Invalid;
+
+        OneByteOfIncompleteMultiByteSequence:
+            numBytesConsumed = 1;
+            return SequenceValidity.Incomplete;
+
+        TwoBytesOfIncompleteMultiByteSequence:
+            numBytesConsumed = 2;
+            return SequenceValidity.Incomplete;
+
+        ThreeBytesOfIncompleteMultiByteSequence:
+            numBytesConsumed = 3;
+            return SequenceValidity.Incomplete;
+        }
+
+        /// <summary>
+        /// We don't have access to Math.DivRem, so this is a copy of the implementation.
```

Comment:
> We don't have access to Math.DivRem

Why not?

Reply:
Artifact of porting for corefxlab which targets netstandard1.1. Will update. However, we will likely have to bring it back anyway in the future.

Reply:
> However, we will likely have to bring it back anyway in the future.

Why?

Reply:
To support a source package that targets netstandard1.x, I will need to ifdef the use of Math.DivRem anyway.

Reply:
Is netstandard1.x a goal? AFAIK netstandard2.0 is the minimum that we're guaranteeing in source.

Reply:
You got a bunch of other problems with turning it into nuget package, like the direct dependency on CoreLib:

https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/src/System.Text.Json.csproj#L36

Also, if you hope to turn this into netstandard nuget package with same name, you should cleanup this CoreLib dependency before we ship. Otherwise, this will become one of those immediately dead partially inboxed NuGet packages that we really want to avoid going forward.

Reply:
> You got a bunch of other problems with turning it into nuget package, like the direct dependency on CoreLib

Yes, I plan to fix the dependencies once we ifdef the code to work on netstandard (will likely need conditional includes in the csproj).

> you should cleanup this CoreLib dependency before we ship

Yes, of course.

> Is netstandard1.x a goal?

I believe it is for .NET Framework support.

cc @KrzysztofCwalina